### PR TITLE
Add metadata_max config to help performance on devices with large blocks

### PR DESCRIPTION
**Pull request by mon**
Alright, hopefully I've done this correctly. Inspired by our conversation on #203.
Not sure of the best way to validate performance "properly", if there's a test bench already setup I should be using, but I am still seeing a 10x increase in directory traversal/file deletion and a 2x-4x speedup in file reads with this code and a `metadata_max` of 4096 and `inline_file_max` of -1.

Commit message:
We have seen poor read performance on NAND flashes with 128kB blocks.
The root cause is inline files having to traverse many sets of metadata
pairs inside the current block before being fully reconstructed. Simply
disabling inline files is not enough, as the metadata will still fill up
the block and eventually need to be compacted.

By allowing configuration of how much size metadata takes up, along with
limiting (or disabling) inline file size, we achieve read performance
improvements on an order of magnitude.

### CODE REVIEW

**Comment by mon** [comment]
Also, as far as backwards compat/superblock inclusion goes, I've been able to move backwards and forwards between custom/default values with no data loss, errors or asserts. I don't know the codebase well enough to be sure, but I believe since these new limits only come into play when deciding to compact, adding new blocks, or rewriting files, they will be silently applied only when required, and old blocks can be traversed fine.

**Comment by geky** [comment]
This looks great!

The only question I have is do we need `inline_file_max` if we have `metadata_max`? It is currently based on `block_size/8`, but could instead be changed to `metadata_max/8`. I'm not sure a separate `inline_file_max` is useful.

File: lfs.h [hunk 1]
```diff
@@ -406,6 +420,8 @@ typedef struct lfs {
     lfs_size_t name_max;
     lfs_size_t file_max;
     lfs_size_t attr_max;
+    lfs_size_t metadata_max;
```
Comment (geky) [inline lfs.h:423]: Ah, if `metadata_max` is not dependent on the superblock, it's better to not keep it in the `lfs_t` struct, but instead calculate its value on every use (`lfs->metadata_max ? lfs->metadata_max : lfs->cfg->block_size`).

Something similar is done for `block_cycles`. It's unfortunately a bit noisy, but it reduces RAM at the cost of code size, and has the potential to be optimized out in https://github.com/littlefs-project/littlefs/pull/491.

**Comment by geky** [comment]
> Not sure of the best way to validate performance "properly", if there's a test bench already setup I should be using

Unfortunately not at the moment. It's on my TODO list to put together a set of benchmarks similar to the tests, and would be the first step in addressing the current performance issues.

Performance hasn't been a high priority for littlefs, and the only real basis for performance during development was theory, so the current performance issues probably shouldn't have been a surprise. I believe "you are what you test" is how the saying goes...

File: lfs.h (hunk 1 above)
Comment (mon) [reply lfs.h:423]: Ah, I was a little worried at how sparse `lfs_t` was and whether this was the right choice. I'll modify this to calculate-at-use when I sort the inline files config.

**Comment by geky** [comment]
> all our files are strictly larger than the inline file size, so I want to be able to disable them entirely

Ah, if you create a file larger than the inline file size, it should never get written out to disk as an inline file (unless there's a bug, or you call `lfs_file_sync` while the file is still inline size).

Due to other conditions with multiple open files, open inline files are stored entirely in the file cache. So in theory, when you write a file larger than cache_size, by the time littlefs needs to flush the data to disk it will have already decided the file should not be inlined.

**Comment by mon** [comment]
Aha, and that explains why we had the perf issues at 128k blocks and 4k cache - the inline file limit is 16k, so it will always be written out before un-inlining.

Fix commits incoming shortly!

**Comment by geky** [comment]
Looks perfect 👍 Thanks for following all the unwritten style rules.

Will bring this in on the next minor release.

**Review by geky — APPROVED** [review]

**Comment by geky** [comment]
Assuming it passes CI ;)

But it did pass the first time

**Comment by geky** [comment]
> perf issues at 128k blocks and 4k cache - the inline file limit is 16k

Hmmm, the actual check is against `min(0x3fe, cache_size, block_size/8)` (0x3fe is due to a 10-bit size field). So in theory it still should not have been written out as inlined:
https://github.com/littlefs-project/littlefs/blob/1a59954ec64ca168828a15242cc6de94ac75f9d1/lfs.c#L2966-L2969

**Comment by mon** [comment]
Hmm, are directories/dir entries stored as files, and thus inline-able? That's the only other possibility if they were indeed being inlined as that 0x3fe shows.

**Comment by geky** [comment]
Nope, littlefs doesn't do directory inlining. I wonder if it might have been a bug.

**Comment by geky** [comment]
Thanks for the PR!

**Pull request merged**